### ISSUE TITLE
How do I install docker via YUM provider, not RPM?

### ISSUE THREAD

**Issue by nick-george** [puppetlabs/puppetlabs-docker#225]
Hi, 

It appears that your code is forcing the installation of docker via RPM, not YUM. Is that correct? I have the 'docker-engine' package in a YUM repository and am having the following error

> Error: Execution of '/bin/rpm -i docker-engine' returned 1: error: open of docker-engine failed: No such file or directory
Error: /Stage[main]/Docker::Install/Package[docker]/ensure: change from absent to present failed: Execution of '/bin/rpm -i docker-engine' returned 1: error: open of docker-engine failed: No such file or directory

I thought I might be able to make use of the repo_opt parameter, but it doesn't look like I can override the package provider to be YUM instead of RPM. 

Can you please explain to me how I can use your module to install docker via YUM? 

Here's my current code 
`
class {'::docker':
    use_upstream_package_source => false,
    package_source                          => 'docker-engine'
}
`

Thanks, 
Nick

**Comment by jmbelvar81** [comment]
Hi Nick,

I'm facing a similar problem. If I try use the official package defining the *use_upstream_package_source* to *false* and *version* to my required number *1.13.1* , the errors obtained are the next one:

  - Error: Could not update: Execution of '/usr/bin/yum -d 0 -e 0 -y --enablerepo=rhel-7-server-extras-rpms install docker-engine-1.13.1' returned 1: Error: Nothing to do

 - Error: /Stage[main]/Docker::Install/Package[docker]/ensure: change from 'purged' to '1.13.1' failed: Could not update: Execution of '/usr/bin/yum -d 0 -e 0 -y --enablerepo=rhel-7-server-extras-rpms install docker-engine-1.13.1' returned 1: Error: Nothing to do

The docker is installed but the service never runs although I configured  next attributes: 

  - service_state : running
  - service_enable: true
  - manage_service: true

My declaration for the class is : 
```
class{'docker':
       ensure                      => 'present',                        
       service_name                => 'docker',                         
       version                     => '1.13.1',                         
       use_upstream_package_source => false ,                           
       selinux_enabled             => true,
       labels                      => ['docker_ose_3_9'],               # Label for the service, completely "arbitrary". 
       service_state               => 'running',
       service_enable              => true,                             # The service was created but "not started" I fear
       manage_service              => true,                             # Without this option the other service options are ignored
    }  

```
In next executions it becomes worst because I obtain the next errors:

  - **Error: Facter: error while resolving custom fact "docker": 751: unexpected token at ''**
  - **Error: Facter: error while resolving custom fact "docker_server_version": undefined method `[]' for nil:NilClass**

But, the strange thing is that I can start the service using systemctl docker start and it runs perfectly :-(

I hope that this info at least help you to can install the service using docker class. 

*Note*: I tried too using the http url that I obtained from yumdb for docker but it doesn't solve the mentioned problems.

**Comment by jmbelvar81** [comment]
Some minutes ago I open a new request to implement the mentioned change in my previous message as the next commit indicates:

  - **commit b9c3cd771d962b797e85deb4ca0cee8d7f92ef45**

Any suggestion is welcome :-)

**Comment by nick-george** [comment]
Thanks @jmbelvar81, I think your PR would work well for me. I think the dev must have chosen RPM as the provider for a reason, but I can't think why. It will be interesting to see if it is excepted. I'm curious why the very first error message you mention in your comment has YUM and not RPM as the installer. 

Cheers, 
Nick

**Comment by jmbelvar81** [comment]
Hi Nick,

Happy to help always if I can. I hope my change could help you to solve your problem finally. 

About your question,  I suppose they are using rpm and dpkg because they are assuming package install (.deb or .rpm) in "manual" way (nor package manager: apt / yum). In my case the described change avoided  the error and yum was used because it's the default package manager on my RHEL 7.x , I think. Maybe it will be needed analyze a greater number of "use case" in relation to install methods that involve package managers and different repositories.

In relation to my first error I think that enable the repository should be an option offered by the module too , because I had enabled the repository "extras" as result of previous processes that registered that host with the subscription manager.

In any way It's very probable that my PR will be not enough to solve each type of errors related to docker package install process, I'm waiting the answer for module dev/s :-) to know the "final" solution from them. 

R,

Jose M

P.S: My apologies for my bad english, It's not my native language

**Comment by LongLiveCHIEF** [comment]
> About your question, I suppose they are using rpm and dpkg because they are assuming package install (.deb or .rpm) in "manual" way (nor package manager: apt / yum).

Actually, they are using these because it's legacy code in the repo. When the original module was written docker was not yet available in either official channels, nor via docker channels. Docker only distributed as `.deb` and `.rpm`.  This code has just never been refactored here.

I keep planning to get around to this, and I _did_ write an update for my own fork, but it's only compatible with Puppet 5.

One of the workarounds i've seen some people do is `manage_package = false` for this module, and then they will use my [repo/package](https://github.com/HackerHappyHour/hackerhappyhour-docker/blob/7-refactor-package-management/manifests/repository.pp) code from [hackerhappyhour-docker/branches/7-refactor-package-management](https://github.com/HackerHappyHour/hackerhappyhour-docker/tree/7-refactor-package-management)

**Comment by nick-george** [comment]
Thanks for your comment @LongLiveCHIEF. I think I'll go with your second option if @jmbelvar81 PR isn't accepted.

**Comment by nick-george** [comment]
> P.S: My apologies for my bad english, It's not my native language

Hi Jose, 

No need to apologise, I am in-awe at how people with non English speaking backgrounds are able to function perfectly well on predominantly English speaking websites. I wish I had taken the time and effort to learn another language, the problem is, there are just too many! 

Cheers, 
Nick

**Comment by scotty-c** [comment]
merged into master #227

**Comment by jmbelvar81** [comment]
Thank you for your answer @LongLiveCHIEF, It's good to know that the reason is "legacy code" in the repo. About your solution , in my case, is not an option because my clients require that only use approved modules and the functionality included in them. If now it's included , of course I will use it immediatly ;-) . It's much more elegant that my "quick fix" :-)

Thanks again for so useful information and, overall, your repo :-)

P.S: Also Thank you Nick for your understanding about language :-) and thanks too for your collaboration in this conversation :-) . I was the "seed" that allow create a new PR and increase our knowledge of this "great" module. The Puppet community members have to help us :-)
